jupyter_kernel_mgmt/subproc/manager.py

```diff
@@ -40,11 +47,20 @@ def __init__(self, popen, files_to_cleanup=None, win_interrupt_evt=None):
         self.win_interrupt_evt = win_interrupt_evt
         self.log = get_app_logger()
         self.kernel_id = str(uuid.uuid4())
-        self._exit_future = asyncio.ensure_future(self.kernel.wait())
+        if not use_sync_subprocess:
+            self._exit_future = asyncio.ensure_future(self.kernel.wait())
 
     async def wait(self):
         """Wait for kernel to terminate"""
-        await self.kernel.wait()
+        if use_sync_subprocess:
+            sync_wait_timeout = 2
+            try:
+                self.kernel.wait(sync_wait_timeout)
+            except subprocess.TimeoutExpired:
+                self.log.warning("Timeout expired waiting for process '{}' to terminate.  Continuing...".
```

Comment:
This seems very ugly - it's impossible to wait for longer than 2 seconds, but the event loop could be blocked for that long. I think the simplest improvement is to implement a polling loop with something like a 0.1 second async sleep in here - then it works basically as expected, just with a tiny extra delay after the process has really terminated.

Reply:
Agreed - I'm disappointed I left it like that, but I did - sorry.  

It sounds like you're talking about implementing something close to the previous [`finish_shutdown()`](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/manager.py#L287-L308) method.  I'll go ahead with this, leaving things essentially as they are in jupyter_client (i.e., 5 second max).

Reply:
Similar implementation, but we don't need any timeout here (because you can impose it from outside for async code). Just something like:

```python
while self.kernel.poll() is None:
    await asyncio.sleep(0.1)
```

Reply:
(And no need to apologise - I think I had also thought this was harder to work around before.)

Reply:
Thanks for the responses. I guess I’m not following how the timeout would be imposed. Could you provide an example?

Reply:
```python
await asyncio.wait_for(manager.wait(), timeout=3.0)
```

The asyncio cancellation mechanism should be able to interrupt `asyncio.sleep()` perfectly fine, but it can't interrupt the blocking `Popen.wait()` call.

Reply:
Ok, but manager.wait() is likely only called by manager.kill() so are you suggesting that this be used from kill() and is 3 seconds sufficient?

Reply:
kill could use something like that - I've no particular preference on what the timeout should be. It's quite unusual for a process to hang so it can't be killed, but it is possible.

But other things could also call wait - e.g. the restarter machinery, or code outside this repo.